src/audio/codec_adapter/codec_adapter.c

```diff
@@ -373,38 +392,73 @@ static void generate_zeroes(struct comp_buffer *sink, uint32_t bytes)
 	comp_update_buffer_produce(sink, bytes);
 }
 
-int codec_adapter_copy(struct comp_dev *dev)
+static void module_copy_samples(struct comp_dev *dev, struct comp_buffer *src_buffer,
+				struct comp_buffer *sink_buffer, uint32_t produced)
 {
-	int ret = 0;
-	uint32_t bytes_to_process, copy_bytes, processed = 0, produced = 0;
-	struct comp_buffer *source = list_first_item(&dev->bsource_list, struct comp_buffer,
-						     sink_list);
-	struct comp_buffer *sink = list_first_item(&dev->bsink_list, struct comp_buffer,
-						    source_list);
 	struct processing_module *mod = comp_get_drvdata(dev);
-	struct module_data *md = &mod->priv;
-	struct comp_buffer *local_buff = mod->local_buff;
 	struct comp_copy_limits cl;
-	struct list_item *blist;
-	size_t size = MAX(mod->deep_buff_bytes, mod->period_bytes);
-	int i = 0;
+	uint32_t copy_bytes;
 
-	if (!source || !sink) {
-		comp_err(dev, "codec_adapter_copy(): source/sink buffer not found");
-		return -EINVAL;
+	if (!produced && !mod->deep_buff_bytes) {
+		comp_dbg(dev, "module_copy_samples(): nothing processed in this call");
+		/*
+		 * No data produced anything in this period but there still be data in the buffer
+		 * to copy to sink
+		 */
+		if (audio_stream_get_avail_bytes(&src_buffer->stream) >= mod->period_bytes)
+			goto copy_period;
+
+		return;
 	}
 
-	comp_dbg(dev, "codec_adapter_copy() start: local_buff free: %d source avail %d",
-		 local_buff->stream.free, source->stream.avail);
+	if (mod->deep_buff_bytes) {
+		if (mod->deep_buff_bytes >= audio_stream_get_avail_bytes(&src_buffer->stream)) {
+			generate_zeroes(sink_buffer, mod->period_bytes);
+			return;
+		}
+
+		comp_dbg(dev, "module_copy_samples(): deep buffering has ended after gathering %d bytes of processed data",
+			 audio_stream_get_avail_bytes(&src_buffer->stream));
+		mod->deep_buff_bytes = 0;
+	}
+
+copy_period:
+	comp_get_copy_limits_with_lock(src_buffer, sink_buffer, &cl);
+	copy_bytes = cl.frames * cl.source_frame_bytes;
```

Comment:
What if copy_bytes is zero?

Reply:
I dont think you can get here is copy_bytes is 0. we get to this point only is audio_stream_get_avail_bytes(&src_buffer->stream) conditions above are met

Reply:
for dai comp, if comsumer(codec) slow than producter(codec_adapter), the dai buffer will be fulled, so in this condition, copy_bytes should be 0.

Reply:
@lenghonglin even if copy_bytes is 0, the operations below will return without doing anything in that case. But sure, I can add a check for this in a follow up PR